scripts/word_embeddings/train_fasttext.py

```diff
@@ -260,106 +290,155 @@ def train(args):
-                                       subwords, subwords_mask).sum(axis=-2)
-
-                    center_negatives = mx.nd.concat(
-                        center.expand_dims(1), negatives, dim=1)
-                    center_negatives_mask = mx.nd.concat(
-                        center_mask.expand_dims(1), mx.nd.ones_like(negatives),
-                        dim=1)
+                    word_context = word_context.reshape((-3, 1))
+                    word_context_mask = word_context_mask.reshape((-3, 1))
+                    if args.ngram_buckets:
+                        emb_in = embedding(word_context, subwords,
+                                           word_context_mask, subwords_mask,
+                                           inverse_unique_indices)
+                    else:
+                        emb_in = embedding(word_context, word_context_mask)
+
+                    with mx.autograd.pause():
+                        center = center.tile(args.window * 2).reshape((-1, 1))
+                        negatives = negatives.reshape((-1, args.negative))
+
+                        center_negatives = mx.nd.concat(
+                            center, negatives, dim=1)
+                        center_negatives_mask = mx.nd.concat(
+                            mx.nd.ones_like(center), negatives_mask, dim=1)
 
                     emb_out = embedding_out(center_negatives,
                                             center_negatives_mask)
 
                     # Compute loss
-                    pred = mx.nd.batch_dot(
-                        emb_in.expand_dims(1), emb_out.swapaxes(1, 2))
-                    pred = pred.reshape((-1, 1 + args.negative))
+                    pred = mx.nd.batch_dot(emb_in, emb_out.swapaxes(1, 2))
+                    pred = pred.squeeze() * word_context_mask
                     label = mx.nd.concat(
-                        mx.nd.ones_like(center).expand_dims(1),
+                        mx.nd.ones_like(word_context),
                         mx.nd.zeros_like(negatives), dim=1)
 
                 loss = loss_function(pred, label)
 
             loss.backward()
 
-            if args.optimizer.lower() not in ['adagrad', 'adam']:
-                trainer.set_learning_rate(args.lr * (1 - progress))
-            if args.optimizer_subwords.lower() not in ['adagrad', 'adam']:
-                trainer_subwords.set_learning_rate(args.lr * (1 - progress))
+            if args.optimizer.lower() != 'adagrad':
+                trainer.set_learning_rate(
+                    max(0.0001, args.lr * (1 - progress)))
+
+            if (args.optimizer_subwords.lower() != 'adagrad'
+                    and args.ngram_buckets):
+                trainer_subwords.set_learning_rate(
+                    max(0.0001, args.lr_subwords * (1 - progress)))
+
             trainer.step(batch_size=1)
-            trainer_subwords.step(batch_size=1)
+            if args.ngram_buckets:
+                trainer_subwords.step(batch_size=1)
```

Comment:
The step size at time t should be , r_0 * (1 −t/TP ) ,where r_0=0.025 and 0.05 for skipgram and cbow respectively.

Reply:
It seems necessary to use an adaptive optimizer for getting good results when training with batching. That's why the step sizes are different.

Reply:
Ie. this implementation uses Adagrad with batching, but the C++ implementation usually use an asynchronous SGD